app/Http/Controllers/Auth/OAuthController.php

```diff
@@ -44,47 +44,84 @@ public function redirect($provider)
      * @param  string  $driver
      * @return \Illuminate\Http\Response
      */
-    public function handleCallback($provider)
+    public function handleCallback(OAuthProviderService $provider)
     {
-        $user = Socialite::driver($provider)->stateless()->user();
-        $user = $this->findOrCreateUser($provider, $user);
-
-        $this->guard()->setToken(
-            $token = $this->guard()->login($user)
-        );
-
-        return view('oauth/callback', [
-            'token' => $token,
-            'token_type' => 'bearer',
-            'expires_in' => $this->guard()->getPayload()->get('exp') - time(),
-        ]);
+        try {
+            $driverUser = $provider->getDriver()->setRedirectUrl(config('services.google.auth_redirect'))->getUser();
+            $user = $this->findOrCreateUser($provider, $driverUser);
+            if(!$user) {
+                return $this->error([
+                    "message" => "User not found."
+                ]);
+            }
+            $this->guard()->setToken(
+                $token = $this->guard()->login($user)
+            );
+
+            return response()->json([
+                'token' => $token,
+                'token_type' => 'bearer',
+                'expires_in' => $this->guard()->getPayload()->get('exp') - time(),
+                'new_user' => $user->new_user
+            ]);
+        } catch(\Exception $e) {
+            return $this->error([
+                "message" => "OAuth service failed to authenticate: ". $e->getMessage()
+            ]);
+        }
     }
 
     /**
-     * @param  string  $provider
-     * @param  \Laravel\Socialite\Contracts\User  $sUser
-     * @return \App\Models\User
+     * @param  \Laravel\Socialite\Contracts\User  $socialiteUser
+     * @return \App\Models\User | null
      */
-    protected function findOrCreateUser($provider, $user)
+    protected function findOrCreateUser($provider, $socialiteUser)
     {
         $oauthProvider = OAuthProvider::where('provider', $provider)
-            ->where('provider_user_id', $user->getId())
+            ->where('provider_user_id', $socialiteUser->getId())
             ->first();
 
         if ($oauthProvider) {
             $oauthProvider->update([
-                'access_token' => $user->token,
-                'refresh_token' => $user->refreshToken,
+                'access_token' => $socialiteUser->token,
+                'refresh_token' => $socialiteUser->refreshToken,
             ]);
 
             return $oauthProvider->user;
         }
 
-        if (User::where('email', $user->getEmail())->exists()) {
-            throw new EmailTakenException();
-        }
 
-        return $this->createUser($provider, $user);
+        if ($provider->getDriver()->canCreateUser()) {
```

Comment:
let's simplify the different IF statements. Instead of creating huge statements, we can negate them. For instance 
```php
if (!$provider->getDriver()->canCreateUser()) { // use negation
  // do something
  return $something;
}

// Rest of the code
```

Also:
- when creating a user: always make sure email is lowercase
- when searching a user by email: use lowercase
- if there's already a user using the same email (but not connected to google account), then respond with an error message explaining that use is already registered with this email and need to use password